archive_repo.py

```diff
@@ -0,0 +1,481 @@
+        if node != None:
+            for key in [*node]:
+                if key not in repo_keys_to_keep:
+                    del node[key]
+
+    commits, etag = retrieve_and_scrub(
+        pr_details["commits_url"],
+        old_pr_details,
+        "etag_commits",
+        "commits",
+        scrub_commits,
+    )
+    pr_details["commits"] = commits
+    pr_details["etag_commits"] = etag
+    del pr_details["commits_url"]
+
+    comments, etag = retrieve_and_scrub(
+        pr_details["comments_url"],
+        old_pr_details,
+        "etag_comments",
+        "comments",
+        scrub_comments,
+    )
+    pr_details["comments"] = comments
+    pr_details["etag_comments"] = etag
+    del pr_details["comments_url"]
+
+    review_comments, etag = retrieve_and_scrub(
+        pr_details["review_comments_url"],
+        old_pr_details,
+        "etag_review_comments",
+        "review_comments",
+        scrub_review_comments,
+    )
+    pr_details["review_comments"] = review_comments
+    pr_details["etag_review_comments"] = etag
+    del pr_details["review_comments_url"]
+
+    return pr_details
+
+
+#####################
+## Body of program ##
+#####################
+
+## Read in the reference files, if any
+issue_ref = dict()
+issues_to_scrub = list()
+pr_ref = dict()
+deleted_issues = dict()
+
+for file_name, to_scrub, dest in (
+    (args.refIssues, issues_to_scrub, issue_ref),
+    (args.refPullRequests, None, pr_ref),
+):
+    if file_name:
+        with open(file_name, "r") as ref_file:
+            raw_reference = json.load(ref_file)
+
+            for element in raw_reference:
+                # If there is no "comments" node, it's not usable
+                if element != None and "comments" in element.keys():
+                    if "number" in element.keys() and not isinstance(
+                        element["comments"], int
+                    ):
+                        dest[element["number"]] = element
+                    else:
+                        # Scrub anything with a number for the "comments" element
+                        # -- not generated by this script
+                        if to_scrub is not None:
+                            to_scrub.append(element)
+
+# Implicit assumption here that a file is all new or all old; mixing gets dicey.
+#
+# So, if we had old input, we have to output everything in the new format:
+if (args.refIssues and not issue_ref) or (args.refPullRequests and not pr_ref):
+    args.omitOld = False
+
+# If we had an old issues file, we can update it to the new format without re-fetching all issues.
+scrub_issues(issues_to_scrub)
+for issue in issues_to_scrub:
+    issue_ref[issue["number"]] = issue
+
+# An old pulls file isn't usable, unfortunately.
+if not pr_ref:
+    args.refPullRequests = None
+
+
+## Download from GitHub the full issues list (if no date) or the updated issues list (if date)
+url = "https://api.github.com/repos/" + args.repo + "/issues?state=all&per_page=100"
```

Comment:
I notice that you are using the v3 API here still.  Did you look at the graphql version at all, as mentioned in #161?  I think that might be better at getting you what you want without a whole lot of post-processing.  I say that because this script is very large and complicated and I'm looking for ways to reduce the amount of code we carry.

I also looked at PyGithub, which seems like it might work very nicely for retrieving data.  It might not work so well without login details or with rate limiting though.  Better than maintaining custom code though.

Reply:
I did look at the v4 version.  I initially passed on it, because I couldn't find a way to do the `&since=` piece.  After some more looking, I *think* I see how to do it.  Like v3, only for issues, so I'll then have to download the PRs one by one.

I think it's going to look fairly similar; number of calls might or might not be smaller.  For example, you still have to paginate the issues, so the same number of calls there.  The first several comments can be returned with the issues; that will be sufficient most of the time, but I'll need to crawl the returned comments list to verify that the comments themselves haven't been paginated.  (The v3 API just returns all comments.)

The main advantage seems to be that I can specify which fields I want (and don't) and have that filtering done on the server versus the client.  The main disadvantage is that it's harder to dynamically generate the necessary pagination query in Python than it is to follow the `_url` members on the v3 objects.

Reply:
Looking further. Using the v3 API, the `/issues` endpoint returned both issues and pull requests.  In the v4 API, the `IssueConnection` supports a `filterBy` argument, enabling the same "issues since X" semantic to be expressed.

However, the `Issues` type appears to actually be only issues, not PRs.  The `PullRequestConnection` does not expose a `filterBy` argument.  For now, it appears that there is no way to retrieve only the updated PRs.  I've e-mailed support to ask if there's something I'm missing or if this is slated for future implementation.